# EP2_2: 와인 분류
* 와인 3가지 종류로 분류하는 문제 
* 13가지 특성을 가진 총 178개의 와인 데이터를 이용.

## 1. 데이터 준비 및 확인 

In [12]:
# (1) 필요한 모듈 import하기
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

In [2]:
# (2) 데이터 준비
wine = load_wine()

In [7]:
print(wine.keys())

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])


In [8]:
# (3) 데이터 이해하기
# Feature Data 지정하기
wine_data = wine.data
print(wine_data.shape)
# Label Data 지정하기
wine_label = wine.target
# Target Names 출력해 보기
print("<Target Names>")
print(wine.target_names)
# 데이터 Describe 해 보기
print("<DESCR>")
print(wine.DESCR)

(178, 13)
<Target Names>
['class_0' 'class_1' 'class_2']
<DESCR>
.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                     

## 2. train, test 데이터 나누기

In [10]:
# (4) train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(wine_data, wine_label, test_size=0.2, random_state=7, stratify=wine_label)

print('X_train 개수:', len(X_train), ', X_test 개수:', len(X_test))
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

X_train 개수: 142 , X_test 개수: 36
(142, 13) (142,)
(36, 13) (36,)


## 3. 모델 학습 및 평가

In [15]:
# (5) 다양한 모델로 학습시켜보기

#  Decision Tree 사용해 보기
from sklearn.tree import DecisionTreeClassifier
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred1 = decision_tree.predict(X_test)
print("Decision Tree: \n", classification_report(y_test, y_pred1))
print(confusion_matrix(y_test, y_pred1))

#  Random Forest 사용해 보기
from sklearn.ensemble import RandomForestClassifier
random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred2 = random_forest.predict(X_test)
print("Random Forest: \n", classification_report(y_test, y_pred2))
print(confusion_matrix(y_test, y_pred2))

#  SVM 사용해 보기
from sklearn import svm
svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred3 = svm_model.predict(X_test)
print("SVM          : \n", classification_report(y_test, y_pred3))
print(confusion_matrix(y_test, y_pred3))

#  SGD Classifier 사용해 보기
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred4 = sgd_model.predict(X_test)
print("SGD Classifier:\n", classification_report(y_test, y_pred4))
print(confusion_matrix(y_test, y_pred4))

#  Logistic Regression 사용해 보기
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)
y_pred5 = logistic_model.predict(X_test)
print("Logistic Reg.: \n", classification_report(y_test, y_pred5))
print(confusion_matrix(y_test, y_pred5))

Decision Tree: 
               precision    recall  f1-score   support

           0       1.00      0.92      0.96        12
           1       0.93      0.93      0.93        14
           2       0.91      1.00      0.95        10

    accuracy                           0.94        36
   macro avg       0.95      0.95      0.95        36
weighted avg       0.95      0.94      0.94        36

[[11  1  0]
 [ 0 13  1]
 [ 0  0 10]]
Random Forest: 
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       1.00      0.93      0.96        14
           2       0.91      1.00      0.95        10

    accuracy                           0.97        36
   macro avg       0.97      0.98      0.97        36
weighted avg       0.97      0.97      0.97        36

[[12  0  0]
 [ 0 13  1]
 [ 0  0 10]]
SVM          : 
               precision    recall  f1-score   support

           0       1.00      0.75      0.86        12
    

/opt/conda/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


## 결과
전반적으로 의사결정나무 모델들의 예측결과가 높게 나왔음. 

Random Forest모델이 가장 높은 정확도(0.98)를 가졌다고 판단함. 

#### 평가지표는 recall을 보는 것이 좋다고 생각함
이유는 1) 전체 샘플의 분포에 영향을 덜 받고, 2) 각 라벨별로 정확한 라벨을 골랐는지를 알 수 있기 때문이다.(recall) 3) 컴퓨터가 예상한 라벨값중에서 맞고 틀린 것에 대한 지표(precision)가 중요한 것이 아님. 